cmd/spire-server/cli/agent/purge.go

```diff
@@ -0,0 +1,156 @@
+	"github.com/spiffe/spire/pkg/common/idutil"
+	"golang.org/x/net/context"
+	"google.golang.org/protobuf/types/known/wrapperspb"
+)
+
+type purgeCommand struct {
+	env        *commoncli.Env
+	expiredFor time.Duration
+	dryRun     bool
+	printer    cliprinter.Printer
+}
+
+func NewPurgeCommand() cli.Command {
+	return NewPurgeCommandWithEnv(commoncli.DefaultEnv)
+}
+
+func NewPurgeCommandWithEnv(env *commoncli.Env) cli.Command {
+	return util.AdaptCommand(env, &purgeCommand{env: env})
+}
+
+func (*purgeCommand) Name() string {
+	return "agent purge"
+}
+
+func (*purgeCommand) Synopsis() string {
+	return "Purge expired agents that were attested using a non-TOFU security model based on a given time"
+}
+
+func (c *purgeCommand) Run(ctx context.Context, _ *commoncli.Env, serverClient util.ServerClient) (err error) {
+	agentClient := serverClient.NewAgentClient()
+	resp, err := agentClient.ListAgents(ctx, &agentv1.ListAgentsRequest{
+		Filter:     &agentv1.ListAgentsRequest_Filter{ByCanReattest: wrapperspb.Bool(true)},
+		OutputMask: &types.AgentMask{X509SvidExpiresAt: true},
+	})
+	if err != nil {
+		return fmt.Errorf("failed to list agents: %w", err)
+	}
+
+	agents := resp.GetAgents()
+	expiredAgents := &expiredAgents{Agents: []*expiredAgent{}}
+
+	for _, agent := range agents {
+		id, err := idutil.IDFromProto(agent.Id)
+		if err != nil {
+			return err
+		}
+
+		expirationTime := time.Unix(agent.X509SvidExpiresAt, 0)
+
+		if time.Since(expirationTime) > c.expiredFor {
+			result := &expiredAgent{AgentID: id}
+
+			if !c.dryRun {
+				if _, err := agentClient.DeleteAgent(ctx, &agentv1.DeleteAgentRequest{Id: agent.Id}); err != nil {
+					result.Error = err.Error()
+				} else {
+					result.Deleted = true
+				}
+			}
+			expiredAgents.Agents = append(expiredAgents.Agents, result)
+		}
+	}
+
+	return c.printer.PrintStruct(expiredAgents)
+}
+
+func (c *purgeCommand) AppendFlags(fs *flag.FlagSet) {
+	fs.DurationVar(&c.expiredFor, "expiredFor", 24*time.Hour, "Amount of time that has passed since the agent's SVID has expired. It is used to determine which agents to purge.")
+	fs.BoolVar(&c.dryRun, "dryRun", false, "Indicates that the command will not perform any action, but will print the agents that would be purged.")
+
+	cliprinter.AppendFlagWithCustomPretty(&c.printer, fs, c.env, c.prettyPrintPurgeResult)
+}
+
+type expiredAgents struct {
+	Agents []*expiredAgent `json:"expired_agents"`
+}
+
+type expiredAgent struct {
+	AgentID spiffeid.ID `json:"agent_id"`
+	Deleted bool        `json:"deleted"`
+	Error   string      `json:"error,omitempty"`
+}
+
+func (c *purgeCommand) prettyPrintPurgeResult(env *commoncli.Env, results ...interface{}) error {
+	if expAgents, ok := results[0].([]interface{})[0].(*expiredAgents); ok {
+		if len(expAgents.Agents) == 0 {
+			env.Println("No agents to purge.")
+			return nil
+		}
+
+		msg := fmt.Sprintf("Found %d expired ", len(expAgents.Agents))
+		msg = util.Pluralizer(msg, "agent", "agents", len(expAgents.Agents))
```

Comment:
❤️

small details like this make a big difference in user perception and experience. thank you 🙌